components/eam/src/physics/crm/samomp/pressure.F90

```diff
@@ -0,0 +1,337 @@
+        xnx=pii/nx_gl
+        xny=pii/ny_gl
+        if(dowally) then
+          jd=j+jt-1
+          facty = 1.d0
+        else
+          jd=(j+jt-0.1)/2.
+          facty = 2.d0
+        endif
+        xj=jd
+        if(dowallx) then
+          id=i+it-1
+          factx = 1.d0
+        else
+          id=(i+it-0.1)/2.
+          factx = 2.d0
+        endif
+        xi=id
+        eign(i,j)=(2.D0*cos(factx*xnx*xi)-2.D0)*ddx2+(2.D0*cos(facty*xny*xj)-2.D0)*ddy2
+      enddo
+    enddo
+
+    !For working aroung PGI OpenACC bug where it didn't create enough gangs
+    numgangs = ceiling(ncrms*(nyp22-jwall)*(nxp2-iwall)/128.D0)
+    !$omp target teams distribute parallel do collapse(3) private(alfa,beta)
+    do j=1,nyp22-jwall
+      do i=1,nxp1-iwall
+        do icrm = 1 , ncrms
+          !$omp flush(alfa,beta)
+          if(dowally) then
+            jd=j+jt-1
+          else
+            jd=(j+jt-0.1D0)/2.D0
+          endif
+          if(dowallx) then
+            id=i+it-1
+          else
+            id=(i+it-0.1D0)/2.D0
+          endif
+          if(id+jd.eq.0) then
+            b=1._8/(eign(i,j)*rho(icrm,1)-a(icrm,1)-c(icrm,1))
+            alfa(1)=-c(icrm,1)*b
+            beta(1)=ff(icrm,i,j,1)*b
+          else
+            b=1._8/(eign(i,j)*rho(icrm,1)-c(icrm,1))
+            alfa(1)=-c(icrm,1)*b
+            beta(1)=ff(icrm,i,j,1)*b
+          endif
+          do k=2,nzm-1
+            e=1._8/(eign(i,j)*rho(icrm,k)-a(icrm,k)-c(icrm,k)+a(icrm,k)*alfa(k-1))
+            alfa(k)=-c(icrm,k)*e
+            beta(k)=(ff(icrm,i,j,k)-a(icrm,k)*beta(k-1))*e
+          enddo
+          ff(icrm,i,j,nzm)=(ff(icrm,i,j,nzm)-a(icrm,nzm)*beta(nzm-1))/(eign(i,j)*rho(icrm,nzm)-a(icrm,nzm)+a(icrm,nzm)*alfa(nzm-1))
+          do k=nzm-1,1,-1
+            ff(icrm,i,j,k)=alfa(k)*ff(icrm,i,j,k+1)+beta(k)
+          enddo
+        enddo
+      enddo
+    enddo
+
+    !-----------------------------------------------------------------
+    n = 0
+    !$omp target teams distribute parallel do collapse(4)
+    do k = 1,nzslab
+      do j = 1,nyp22-jwall
+        do i = 1,nxp1-iwall
+          do icrm = 1 , ncrms
+            f(icrm,i,j,k) = ff(icrm,i,j,k)
+          enddo
+        enddo
+      enddo
+    enddo
+
+    !-------------------------------------------------
+    !   Perform inverse Fourier transformation:
+    if(RUN3D) then
+      !$omp target teams distribute collapse(2)
+      do k=1,nzslab
+        do i = 1 , nx_gl+1
+          !$omp parallel do private(ftmp_y,work)
+          do icrm = 1 , ncrms
+            !$omp flush(ftmp_y,work)
+            ftmp_y = f(icrm,i,:,k)
+            call fft991_crm(ftmp_y,work,trigxj,ifaxj,1,nx2,ny_gl,1,+1)
+            f(icrm,i,:,k) = ftmp_y
+          enddo
+        enddo
+      enddo
+    endif
+    !$omp target teams distribute collapse(2)
+    do k=1,nzslab
+      do j = 1 , ny_gl
+        !$omp parallel do private(ftmp_x, work)
+        do icrm = 1 , ncrms
+          !$omp flush(ftmp_x,work)
```

Comment:
What does the flush directive do?

Reply:
It flushes the cache, and makes a thread’s temporary view of memory consistent with memory and enforces an order on the memory operations of the variables explicitly specified or implied

Reply:
I'd be curious to know what this maps to in CUDA. Any ideas?

Reply:
We may get some ideas if we study the ptx code, actually Eric did this study.